### PR TITLE
Remove dependency on `ruff_rowan`

### PR DESCRIPTION
**Pull request by charliermarsh**
This PR removes the dependency on `ruff_rowan` (i.e., Rome's fork of rust-analyzer's `rowan`), and in turn, trims out a lot of code in `ruff_formatter` that isn't necessary (or isn't _yet_ necessary) to power the autoformatter.

We may end up pulling some of this back in -- TBD. For example, the autoformatter has its own comment representation right now, but we may eventually want to use the `comments.rs` data structures defined in `rome_formatter`.

### CODE REVIEW

**Comment by charliermarsh** [comment]
\cc @MichaReiser

File: crates/ruff_formatter/src/format_extensions.rs [hunk 1]
```diff
@@ -1,178 +0,0 @@
...
-    {
-        Memoized::new(self)
-    }
-}
-
-impl<T, Context> MemoizeFormat<Context> for T where T: Format<Context> {}
-
-/// Memoizes the output of its inner [Format] to avoid re-formatting a potential expensive object.
-#[derive(Debug)]
-pub struct Memoized<F, Context> {
```
Comment (MichaReiser) [inline crates/ruff_formatter/src/format_extensions.rs:71]: You probably want to keep this around. Memoization is essential to get good formatting performance when using the `BestFitting` IR

File: crates/ruff_formatter/src/lib.rs [hunk 2]
```diff
@@ -220,13 +202,6 @@ pub trait FormatContext {
 
     /// Returns the formatting options
     fn options(&self) -> &Self::Options;
-
-    /// Returns [None] if the CST has not been pre-processed.
-    ///
-    /// Returns [Some] if the CST has been pre-processed to simplify formatting.
-    /// The source map can be used to map positions of the formatted nodes back to their original
-    /// source locations or to resolve the source text.
-    fn source_map(&self) -> Option<&TransformSourceMap>;
```
Comment (MichaReiser) [inline crates/ruff_formatter/src/lib.rs:229]: We may need source maps in the long term to enable range formatting (formatting the selected text) or to suppress formatting of specific nodes

File: crates/ruff_formatter/src/token/number.rs [hunk 3]
```diff
@@ -1,296 +0,0 @@
-use crate::token::string::ToAsciiLowercaseCow;
-use ruff_rowan::{Language, SyntaxToken};
-use std::borrow::Cow;
-use std::num::NonZeroUsize;
-
-use crate::prelude::*;
-use crate::{CstFormatContext, Format};
-
-pub fn format_number_token<L>(token: &SyntaxToken<L>) -> CleanedNumberLiteralText<L>
```
Comment (MichaReiser) [inline crates/ruff_formatter/src/token/number.rs:9]: We may use some of this (together with `string`) to normalize the string/number content. But we can pull this in when needed.

File: crates/ruff_formatter/Cargo.toml [hunk 4]
```diff
@@ -11,7 +11,7 @@ cfg-if = { version = "1.0.0" }
 countme = { version = "3.0.1" }
 drop_bomb = { version = "0.1.5" }
 indexmap = { version = "1.9.1" }
```
Comment (MichaReiser) [inline crates/ruff_formatter/Cargo.toml:13]: Did you verify if all these dependencies are still needed? https://github.com/est31/cargo-udeps

**Review by MichaReiser — COMMENTED** [review]
LGTM

File: crates/ruff_formatter/src/token/number.rs (hunk 3 above)
Comment (charliermarsh) [reply crates/ruff_formatter/src/token/number.rs:9]: Yeah we'll absolutely need this kind of logic. But does it belong here, or in `ruff_python_formatter`? Black has rules around this stuff, but some of it is Python-specific (e.g., for complex numbers).

File: crates/ruff_formatter/Cargo.toml (hunk 4 above)
Comment (charliermarsh) [reply crates/ruff_formatter/Cargo.toml:11]: Was able to remove a handful of them -- good call.

**Pull request merged**